lib/utils/prompt/context_reader.go

```diff
@@ -0,0 +1,306 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prompt
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/term"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ErrReaderClosed is returned from ContextReader.ReadContext after it is
+// closed.
+var ErrReaderClosed = errors.New("ContextReader has been closed")
+
+// ErrNotTerminal is returned by password reads attempted in non-terminal
+// readers.
+var ErrNotTerminal = errors.New("underlying reader is not a terminal")
+
+type readOutcome struct {
+	value []byte
+	err   error
+}
+
+type readerState int
+
+const (
+	readerStateIdle readerState = iota
+	readerStateClean
+	readerStatePassword
+	readerStateClosed
+)
+
+// xTermI aggregates methods from /x/term for easy mocking.
+type xTermI interface {
```

Comment:
Perhaps `xTermI` could be `terminalI` and `xTerm` could be made to be `fdTerminal` and it could newtype over `int` to contain the file descriptor or maybe over `*os.File` after verifying that it's a tty? The methods could then be made to work on the wrapped file rather than requiring an external file descriptor.

Reply:
@zmb3 `term` shadows the `term` package, that's why I avoided it.

@espadolini let me take a stab at that.

Reply:
Done a few renames, but I'd rather keep the signatures identical to /x/term methods. Encapsulating the fd number makes it deviate further from how the real thing looks.

Reply:
Fair enough 👍